src/python_testing/TC_DEM_2_2.py

```diff
@@ -0,0 +1,325 @@
+
+        self.step("14")
+        await self.send_power_adjustment_command(power=max_power,
+                                                 duration=max_duration,
+                                                 cause=Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization,
+                                                 expected_status=Status.ConstraintError)
+
+        self.step("15")
+        await self.send_test_event_trigger_user_opt_out_grid()
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.PowerAdjustEnd)
+        asserts.assert_equal(event_data.cause, Clusters.DeviceEnergyManagement.Enums.CauseEnum.kUserOptOut)
+
+        elapsed = datetime.datetime.now() - start
+        asserts.assert_less_equal(abs(elapsed.seconds - event_data.duration), 1)
+
+        # TODO: Do a better check on valid energyUse value.
+        # Value returned here is defined in DeviceEnergyManagementManufacturerImpl::GetEnergyUse()
+        asserts.assert_greater_equal(event_data.energyUse, 300)
+
+        self.step("15a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("15b")
+        await self.check_dem_attribute("OptOutState", Clusters.DeviceEnergyManagement.Enums.OptOutStateEnum.kOptOut)
+
+        self.step("15c")
+        powerAdjustCapabilityStruct = await self.read_dem_attribute_expect_success(attribute="PowerAdjustmentCapability")
+        asserts.assert_equal(powerAdjustCapabilityStruct.cause,
+                             Clusters.DeviceEnergyManagement.Enums.PowerAdjustReasonEnum.kNoAdjustment)
+
+        self.step("16")
+        await self.send_test_event_trigger_user_opt_out_clear_all()
+
+        self.step("16a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("16b")
+        await self.check_dem_attribute("OptOutState", Clusters.DeviceEnergyManagement.Enums.OptOutStateEnum.kNoOptOut)
+
+        self.step("17")
+        await self.send_power_adjustment_command(power=max_power,
+                                                 duration=min_duration,
+                                                 cause=Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization,
+                                                 expected_status=Status.Success)
+
+        self.step("17a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kPowerAdjustActive)
+
+        self.step("17b")
+        powerAdjustCapabilityStruct = await self.read_dem_attribute_expect_success(attribute="PowerAdjustmentCapability")
+        asserts.assert_equal(powerAdjustCapabilityStruct.cause,
+                             Clusters.DeviceEnergyManagement.Enums.PowerAdjustReasonEnum.kLocalOptimizationAdjustment)
+
+        self.step("18")
+        time.sleep(10)
+
```

Comment:
Missing - check that PowerAdjustEnd Event is sent

Reply:
Done